### PR TITLE
Added support to bind to open file descriptors.

### LINKED ISSUE #1107: [Q] Bind to a file descriptor?
Hi guys,

I guess whether it's possible to bind gunicorn to a file descriptor of the open socket? I looked at the code and it seems that there is some code that does this in `sock.py` but I'm not sure whether it's only for child workers purposes or I can specify that in `--bind` parameter somehow.

Thanks!

### PR DESCRIPTION
**Pull request by apollo13**
This is an initial try to fix #1107.

@benoitc Which kind of tests do you expect here? I couldn't find much in the testsuite; I'll add parsing tests for `fd://` but is there anything else I should add?

This commit currently vendors https://github.com/tiran/socketfromfd/blob/master/socketfromfd.py from @tiran -- Christian: Would you be okay with this approach (and relicensing under the gunicorn license) or would you like use to use the pypi package (I'd rather vendor given that gunicorn has no dependency).

### CODE REVIEW

**Comment by apollo13** [comment]
Christian contacted me on IRC, he is okay with vendoring but would prefer the license as is. This shouldn't be a problem since Apache and MIT are compatible.

File: docs/source/settings.rst [hunk 1]
```diff
@@ -55,8 +55,8 @@ bind
 
 The socket to bind.
 
-A string of the form: ``HOST``, ``HOST:PORT``, ``unix:PATH``. An IP is
-a valid ``HOST``.
+A string of the form: ``HOST``, ``HOST:PORT``, ``unix:PATH``, ``fd://FD``. An 
+IP is a valid ``HOST``.
```
Comment (berkerpeksag) [inline docs/source/settings.rst]: Could you make this change in `gunicorn/config.py` instead, regenerate `docs/source/settings.rst` by running `make -C docs html`?

Comment (berkerpeksag) [inline docs/source/settings.rst]: And please add a `versionchanged` marker. I think the new feature release will be 19.9.

File: gunicorn/sock.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 
 from gunicorn import util
 from gunicorn.six import string_types
+from gunicorn.socketfromfd import fromfd
```
Comment (berkerpeksag) [inline gunicorn/sock.py]: I know you've followed the current format, but I wonder if we should rename the module to `_socketfromfd`. We are going to remove `argparse_compat.py`, `six.py` and `selectors.py` in 20 and `socketfromfd` will probably be our only vendored dependency (and if we decide not to support Python 3.3 we can remove `socketfromfd` as well)

Comment (apollo13) [reply gunicorn/sock.py]: @berkerpeksag Why can we remove `socketfromfd` if Py 3.3 is dropped? `socketfromfd` provides functionality which is not available up to Py 3.7 (namely deducing the socket family/type from the fd)

**Comment by apollo13** [comment]
As for the CLI tests, there are currently none for `tcp/unix` either; so I am not sure how that would look like.

File: gunicorn/sock.py (hunk 2 above)
Comment (berkerpeksag) [reply gunicorn/sock.py]: Good point, I only checked when `fromfd` was added to the stdlib.

**Comment by berkerpeksag** [comment]
I was talking about adding something like (untested):

```py
def test_bind_fd():
    with AltArgs(["prog_name", "-b", "fd://42"]):
        app = NoConfigApp()
    assert app.cfg.bind == ["fd://42"]
```

File: docs/source/settings.rst [hunk 3]
```diff
@@ -42,591 +42,86 @@ application specific configuration.
...
-Exceeding this number results in the client getting an error when
-attempting to connect. It should only affect servers under significant
-load.
-
-Must be a positive integer. Generally set in the 64-2048 range.
-
-Worker Processes
-----------------
-
-.. _workers:
```
Comment (berkerpeksag) [inline docs/source/settings.rst]: It would be great if you could only include changes to the bind setting. I can resort and regenerate this in a separate commit.

File: gunicorn/config.py [hunk 4]
```diff
@@ -611,11 +614,11 @@ class WorkerClass(Setting):
         A string referring to one of the following bundled classes:
 
         * ``sync``
-        * ``eventlet`` - Requires eventlet >= 0.9.7 (or install it via 
+        * ``eventlet`` - Requires eventlet >= 0.9.7 (or install it via
```
Comment (berkerpeksag) [inline gunicorn/config.py]: We can leave whitespace fixes to another PR.

**Comment by apollo13** [comment]
Ok, added the test and fixed the whitespace changes/rst files.

**Comment by apollo13** [comment]
Is there anything else you need me to do?

File: gunicorn/util.py [hunk 5]
```diff
@@ -251,6 +251,13 @@ def parse_address(netloc, default_port=8000):
     if re.match(r'unix:(//)?', netloc):
         return re.split(r'unix:(//)?', netloc)[-1]
 
+    if netloc.startswith("fd://"):
+        fd = netloc[5:]
+        try:
+            return int(fd)
+        except ValueError:
+            raise RuntimeError("%r is not a valid file descriptor." % fd)
```
Comment (berkerpeksag) [inline gunicorn/util.py]: master is now Python 3-only and I think we can safely use `raise exc from None` pattern here.

File: gunicorn/config.py [hunk 6]
```diff
@@ -535,8 +535,11 @@ class Bind(Setting):
     desc = """\
         The socket to bind.
 
-        A string of the form: ``HOST``, ``HOST:PORT``, ``unix:PATH``. An IP is
-        a valid ``HOST``.
+        A string of the form: ``HOST``, ``HOST:PORT``, ``unix:PATH``,
+        ``fd://FD`. An IP is a valid ``HOST``.
+
+        .. versionchanged:: 19.10
```
Comment (berkerpeksag) [inline gunicorn/config.py]: master is now targets 20.0, so I think this needs to be 20.0 instead of 19.10. We can discuss whether we should backport this to the 19.x branch later.

File: gunicorn/sock.py (hunk 2 above)
Comment (berkerpeksag) [reply gunicorn/sock.py]: @apollo13 could you add a comment in `gunicorn/socketfromfd.py` to explain that socketfromfd is needed for Python versions between 3.3 and 3.6 so we won't accidentally remove it in the future.

File: gunicorn/config.py [hunk 7]
```diff
@@ -535,8 +535,11 @@ class Bind(Setting):
     desc = """\
         The socket to bind.
 
-        A string of the form: ``HOST``, ``HOST:PORT``, ``unix:PATH``. An IP is
-        a valid ``HOST``.
+        A string of the form: ``HOST``, ``HOST:PORT``, ``unix:PATH``,
+        ``fd://FD`. An IP is a valid ``HOST``.
```
Comment (berkerpeksag) [inline gunicorn/config.py]: Typo: \`\`fd://FD\` (missing a second backtick)

**Comment by berkerpeksag** [comment]
Thanks!
